### ISSUE TITLE
Wrong log_softmax output on cuda device float64 torch>=2.4.1

### ISSUE THREAD

**Issue by jchacks** [pytorch/pytorch#143644]
### 🐛 Describe the bug

I checked the other issues but not sure if they are fully related:
https://github.com/pytorch/pytorch/issues/140222

When applying `F.log_softmax` (and taking the exponential) or `F.softmax` over a large-ish dimension 517 (the size matters) on a cuda device (1080 Ti) using f64 dtype, the result does not sum to 1.  I do not have access to another GPU to test if its specific to the 1080 Ti.

To reproduce
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import sys
import platform

if not torch.cuda.is_available():
    raise RuntimeError("This issue only happens on gpu")


def print_system_info():
    print("\nSystem Information:")
    print(f"Python version: {sys.version}")
    print(f"PyTorch version: {torch.__version__}")
    print(f"CUDA available: {torch.cuda.is_available()}")
    print(f"CUDA version: {torch.version.cuda}")
    print(f"GPU: {torch.cuda.get_device_name()}")
    print(f"Platform: {platform.platform()}")


class TinyModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc = nn.Linear(5, 1)

    def forward(self, x):
        return self.fc(x).squeeze(dim=-1)


def main(device: str, dtype: torch.dtype, dimension_size: int = 517):
    print(f"TESTING: {device} {dtype}")
    device = torch.device(device)
    model = TinyModel().to(device).to(dtype)
    optimizer = optim.Adam(model.parameters())

    x = (torch.randn((5, dimension_size, 5), dtype=dtype, device=device) - 0.5) * 2

    optimizer.zero_grad()
    logits = model(x)

    print("\nLogits info:")
    print(f"Shape: {logits.shape}")
    print(f"Device: {logits.device}")
    print(f"Dtype: {logits.dtype}")
    print(f"Requires grad: {logits.requires_grad}")

    # Test different softmax approaches
    print("\nSoftmax sums:")
    print("Method 1:", F.log_softmax(logits, dim=1).exp().sum(dim=1))

    # Try CPU version
    cpu_logits = logits.cpu()
    print("CPU version:", F.log_softmax(cpu_logits, dim=1).exp().sum(dim=1))

    log_probs = logits - torch.logsumexp(logits, dim=1, keepdim=True)
    print("Method 2:", log_probs.exp().sum(dim=1))


if __name__ == "__main__":
    print_system_info()
    # On my setup 517 was the minimum that caused the issue might be different on different machines
    dim_size = 517
    main("cuda", torch.float64, dim_size)   # Doesnt work
    # main("cuda", torch.float32, dim_size) # Works
    # main("cpu", torch.float64, dim_size)  # Works
    # main("cpu", torch.float32, dim_size)  # Works
```

Outputs (I also tested on cu121):
```
System Information:
Python version: 3.10.15 (main, Oct  8 2024, 00:25:34) [Clang 18.1.8 ]
PyTorch version: 2.5.1+cu118 
CUDA available: True
CUDA version: 11.8
GPU: NVIDIA GeForce GTX 1080 Ti
Platform: Linux-6.8.0-50-generic-x86_64-with-glibc2.39
TESTING: cuda torch.float64

Logits info:
Shape: torch.Size([5, 517])
Device: cuda:0
Dtype: torch.float64
Requires grad: True

Softmax sums:
Method 1: tensor([1.0000, 0.6673, 0.9679, 0.6716, 0.9772], device='cuda:0',
       dtype=torch.float64, grad_fn=<SumBackward1>)
CPU version: tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000], dtype=torch.float64,
       grad_fn=<SumBackward1>)
Method 2: tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000], device='cuda:0',
       dtype=torch.float64, grad_fn=<SumBackward1>)
```

Note the first output the log_softmax.exp does not sum to 1. 
When running with the other combinations cuda float32, cpu float64 and cpu float32 the issue does not appear...

I did try with different torch versions and the issue did not appear in 2.3.1, but was present in 2.4.1.
2.3.1 output:
```
System Information:
Python version: 3.10.15 (main, Oct  8 2024, 00:25:34) [Clang 18.1.8 ]
PyTorch version: 2.3.1+cu121
CUDA available: True
CUDA version: 12.1
GPU: NVIDIA GeForce GTX 1080 Ti
Platform: Linux-6.8.0-50-generic-x86_64-with-glibc2.39
TESTING: cuda torch.float64

Logits info:
Shape: torch.Size([5, 517])
Device: cuda:0
Dtype: torch.float64
Requires grad: True

Softmax sums:
Method 1: tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000], device='cuda:0',
       dtype=torch.float64, grad_fn=<SumBackward1>)
CPU version: tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000], dtype=torch.float64,
       grad_fn=<SumBackward1>)
Method 2: tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000], device='cuda:0',
       dtype=torch.float64, grad_fn=<SumBackward1>)
```
2.4.1 output:
```
System Information:
Python version: 3.10.15 (main, Oct  8 2024, 00:25:34) [Clang 18.1.8 ]
PyTorch version: 2.4.1+cu121
CUDA available: True
CUDA version: 12.1
GPU: NVIDIA GeForce GTX 1080 Ti
Platform: Linux-6.8.0-50-generic-x86_64-with-glibc2.39
TESTING: cuda torch.float64

Logits info:
Shape: torch.Size([5, 517])
Device: cuda:0
Dtype: torch.float64
Requires grad: True

Softmax sums:
Method 1: tensor([1.0000, 0.7082, 0.9840, 0.7065, 0.9930], device='cuda:0',
       dtype=torch.float64, grad_fn=<SumBackward1>)
CPU version: tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000], dtype=torch.float64,
       grad_fn=<SumBackward1>)
Method 2: tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000], device='cuda:0',
       dtype=torch.float64, grad_fn=<SumBackward1>)
```

### Versions

Collecting environment information...
PyTorch version: 2.5.1+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.10.15 (main, Oct  8 2024, 00:25:34) [Clang 18.1.8 ] (64-bit runtime)
Python platform: Linux-6.8.0-50-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080 Ti
Nvidia driver version: 565.57.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.5.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.7
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               32
On-line CPU(s) list:                  0-31
Vendor ID:                            AuthenticAMD
Model name:                           AMD Ryzen 9 7950X3D 16-Core Processor
CPU family:                           25
Model:                                97
Thread(s) per core:                   2
Core(s) per socket:                   16
Socket(s):                            1
Stepping:                             2
CPU(s) scaling MHz:                   59%
CPU max MHz:                          5759.0000
CPU min MHz:                          545.0000
BogoMIPS:                             8383.82
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid overflow_recov succor smca fsrm flush_l1d
Virtualisation:                       AMD-V
L1d cache:                            512 KiB (16 instances)
L1i cache:                            512 KiB (16 instances)
L2 cache:                             16 MiB (16 instances)
L3 cache:                             128 MiB (2 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-31
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Mitigation; Safe RET
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] mypy==1.13.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.3
[pip3] nvidia-cublas-cu11==11.11.3.6
[pip3] nvidia-cuda-cupti-cu11==11.8.87
[pip3] nvidia-cuda-nvrtc-cu11==11.8.89
[pip3] nvidia-cuda-runtime-cu11==11.8.89
[pip3] nvidia-cudnn-cu11==9.1.0.70
[pip3] nvidia-cufft-cu11==10.9.0.58
[pip3] nvidia-curand-cu11==10.3.0.86
[pip3] nvidia-cusolver-cu11==11.4.1.48
[pip3] nvidia-cusparse-cu11==11.7.5.86
[pip3] nvidia-nccl-cu11==2.21.5
[pip3] nvidia-nvtx-cu11==11.8.86
[pip3] torch==2.5.1+cu118
[pip3] triton==3.1.0
[conda] Could not collect

cc @ptrblck @msaroufim @eqy

**Comment by malfet** [comment]
To the best of my knowledge, lot of GeForce cards do not have real fp64 hardware, but I can repro the issue on A100:
```
System Information:
Python version: 3.9.5 (default, Nov 23 2021, 15:27:38) 
[GCC 9.3.0]
PyTorch version: 2.6.0a0+gitb26a255
CUDA available: True
CUDA version: 12.3
GPU: NVIDIA A100-SXM4-40GB
Platform: Linux-5.15.0-1019-aws-x86_64-with-glibc2.31
TESTING: cuda torch.float64

Logits info:
Shape: torch.Size([5, 517])
Device: cuda:0
Dtype: torch.float64
Requires grad: True

Softmax sums:
Method 1: tensor([1.0000, 0.7051, 1.0000, 0.7019, 1.0000], device='cuda:0',
       dtype=torch.float64, grad_fn=<SumBackward1>)
CPU version: tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000], dtype=torch.float64,
       grad_fn=<SumBackward1>)
Method 2: tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000], device='cuda:0',
       dtype=torch.float64, grad_fn=<SumBackward1>)
```

**Comment by eqy** [comment]
I believe this is due to `size` going potentially negative in the `ilpReduce` kernel used by softmax, opened a PR here: https://github.com/pytorch/pytorch/pull/144009

**Comment by albanD** [comment]
@eqy can you confirm that the PR above that was merged fixes this issue?

**Comment by eqy** [comment]
Yes, it should and the test that was added is basically from this repro.
